### PR TITLE
Adds parsing of OpenLr type PointAlongLine

### PR DESCRIPTION
**Pull request by PureW**

### CODE REVIEW

File: test/openlr.cc [hunk 1]
```diff
@@ -39,38 +39,46 @@ struct testfixture {
...
+      0, Orientation::NoOrientation, SideOfTheRoad::DirectlyOnRoadOrNotApplicable},
+     {"CwRbWyNG9RpsCQCb/jsbtAT/6/+jK1lE", 6.1268198, 49.6085178, 140.625, 6.1281598, 49.6030578,
+      286.875, 527.4, 68, 0, Orientation::NoOrientation,
+      SideOfTheRoad::DirectlyOnRoadOrNotApplicable},
+     {"CwB67CGukRxiCACyAbwaMXU=", 0.6752192, 47.3651611, 28.125, 0.6769992, 47.3696011, 196.875,
+      468.8, 0, 117, Orientation::NoOrientation, SideOfTheRoad::DirectlyOnRoadOrNotApplicable},
+     {"CwcX6CItqAs6AQAAAAALGg==", 9.9750602, 48.0632865, 298.125, 9.9750602, 48.0632865, 298.125,
+      58.6, 0, 0, Orientation::NoOrientation, SideOfTheRoad::DirectlyOnRoadOrNotApplicable},
+     {"CwRbWyNG9BpgAACa/jsboAD/6/+kKwA=", 6.1268198, 49.6084964, 5.625, 6.1281498, 49.6030464, 5.625,
+      0, 0, 0, Orientation::NoOrientation, SideOfTheRoad::DirectlyOnRoadOrNotApplicable},
```
Comment (PureW) [inline test/openlr.cc:70]: Whitespace change and adding the new `Orientation` and `SideOfTheRoad` property above.

New PointAlongLine tests below

File: src/baldr/openlr.cc [hunk 2]
```diff
@@ -60,7 +60,7 @@ std::vector<std::string> openlr_edges(const TripLeg& leg) {
                       lfrcnp);
     bearing_deg = fmod(bearing_deg + 180., 360.);
     lrps.emplace_back(end.lng(), end.lat(), bearing_deg, frc, fow, &lrps.back());
-    openlrs.emplace_back(OpenLR::LineLocation{lrps, 0, 0}.toBase64());
+    openlrs.emplace_back(OpenLR::OpenLr{lrps, 0, 0}.toBase64());
```
Comment (PureW) [inline src/baldr/openlr.cc:63]: Renaming to `OpenLr` since it now handles point locations as well.

File: valhalla/baldr/openlr.h [hunk 3]
```diff
@@ -138,8 +138,8 @@ struct LocationReferencePoint {
         latitude(!prev ? integer2decimal(decimal2integer(latitude))
                        : prev->latitude +
                              (std::round((latitude - prev->latitude) * geom_scale) / geom_scale)),
-        bearing(integer2bearing(bearing2integer(bearing))), frc(frc), fow(fow),
-        distance(integer2distance(distance2integer(distance))), lfrcnp(lfrcnp) {
+        bearing(integer2bearing(bearing2integer(bearing))),
+        distance(integer2distance(distance2integer(distance))), frc(frc), lfrcnp(lfrcnp), fow(fow) {
```
Comment (PureW) [inline valhalla/baldr/openlr.h:142]: changing the order of property construction to match order in struct (was a warning from compiler).

File: valhalla/midgard/pointll.h [hunk 4]
```diff
@@ -14,7 +14,7 @@ namespace {
 constexpr double RAD_PER_DEG = valhalla::midgard::kPiDouble / 180.0;
 constexpr double DEG_PER_RAD = 180.0 / valhalla::midgard::kPiDouble;
 } // namespace
-constexpr float INVALID_LL = 0xBADBADBAD;
+constexpr float INVALID_LL = (float)0xBADBADBAD;
```
Comment (PureW) [inline valhalla/midgard/pointll.h:17]: Fixing a warning about changed value due to implicit type conversion.

File: valhalla/midgard/pointll.h [hunk 5]
```diff
@@ -37,6 +37,8 @@ template <typename PrecisionT> class GeoPoint : public PointXY<PrecisionT> {
    */
   GeoPoint() : PointXY<PrecisionT>(INVALID_LL, INVALID_LL) {
   }
+  virtual ~GeoPoint() {
```
Comment (PureW) [inline valhalla/midgard/pointll.h:40]: Fixing a warning about non-virtual constructor.

File: valhalla/midgard/pointll.h [hunk 6]
```diff
@@ -208,6 +210,7 @@ template <typename PrecisionT> class GeoPoint : public PointXY<PrecisionT> {
   virtual value_type IsLeft(const GeoPoint& p1, const GeoPoint& p2) const {
     return (p2.lng() - p1.lng()) * (lat() - p1.lat()) - (lng() - p1.lng()) * (p2.lat() - p1.lat());
   }
+  using PointXY<PrecisionT>::IsLeft;
```
Comment (PureW) [inline valhalla/midgard/pointll.h:213]: Fixing a warning about overriding function from parent.

File: valhalla/thor/astar_bss.h [hunk 7]
```diff
@@ -60,7 +60,7 @@ class AStarBSSAlgorithm : public PathAlgorithm {
   /**
    * Clear the temporary information generated during path construction.
    */
-  virtual void Clear();
+  virtual void Clear() override;
```
Comment (PureW) [inline valhalla/thor/astar_bss.h:63]: Fixing a warning.

File: valhalla/baldr/openlr.h [hunk 8]
```diff
@@ -161,25 +161,117 @@ struct LocationReferencePoint {
...
-    }
 
-    const auto raw = reinterpret_cast<const unsigned char*>(decoded.data());
+    const auto raw = reinterpret_cast<const uint8_t*>(decoded.data());
     std::size_t index = 0;
 
     // Status, version 3, has attributes, ArF 'Circle or no area location'
-    auto status = raw[index++] & 0x7f;
-    if (status != 0x0b)
+    const OpenLrStatus status = *reinterpret_cast<const OpenLrStatus*>(&raw[index++]);
```
Comment (kevinkreiser) [inline valhalla/baldr/openlr.h:250]: you dont really need to make a copy of this, you can just as well have a const reference to it. just need to add the `&` to the declaration

File: valhalla/baldr/openlr.h [hunk 9]
```diff
@@ -190,6 +282,13 @@ struct LineLocation {
 
     lrps.emplace_back(longitude, latitude, attribute1, attribute2, attribute3);
 
+    if (isPointAlongLine) {
+      const Attribute5 attr5 = *reinterpret_cast<const Attribute5*>(&attribute1);
```
Comment (kevinkreiser) [inline valhalla/baldr/openlr.h:286]: same criticism as above, no need to copy

File: valhalla/baldr/openlr.h [hunk 10]
```diff
@@ -161,25 +161,117 @@ struct LocationReferencePoint {
...
+  NoOrientation = 0,
+  FirstLrpTowardsSecond = 1,
+  SecondLrpTowardsFirst = 2,
+  BothDirections = 3,
+};
+
+const uint8_t OPENLR_VERSION = 3;
+
+// The first byte of the OpenLr identifier, page 55
+struct OpenLrStatus {
```
Comment (kevinkreiser) [inline valhalla/baldr/openlr.h:191]: thanks for structifying this! we should do the other attributes at some point but this is a great start!

**Review by kevinkreiser — DISMISSED** [review]
minor nits, othe rwise looks good

File: CHANGELOG.md [hunk 11]
```diff
@@ -149,6 +149,8 @@
    * CHANGED: Refactor mapmatching configuration to use a struct (instead of `boost::property_tree::ptree`). [#2485](https://github.com/valhalla/valhalla/pull/2485)
    * ADDED: Save exit maneuver's begin heading when combining enter & exit roundabout maneuvers. [#2554](https://github.com/valhalla/valhalla/pull/2554)
    * ADDED: Added new urban flag that can be set if edge is within city boundaries to data processing; new use_urban_tag config option; added to osrm response within intersections. [#2522](https://github.com/valhalla/valhalla/pull/2522)
+   * ADDED: Parses OpenLr of type PointAlongLine [#2565](https://github.com/valhalla/valhalla/pull/2565)
+
```
Comment (kevinkreiser) [inline CHANGELOG.md:153]: whitespace typo?

Comment (PureW) [reply CHANGELOG.md:153]: I suspect this was a conflict that was automatically resolved by the union merge. Fixing.

**Review by kevinkreiser — APPROVED** [review]

**Pull request merged**